### PR TITLE
Removed squaring from constrained smoother cost function shape

### PR DESCRIPTION
**Pull request by videh25**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #5072 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Loopback Sim |
| Does this PR contain AI generated software? | No |
| Was this PR description generated by AI software? | No |

---

## Description of contribution in a few bullet points

* Removed squaring of residuals in ceres cost fucntions of `nav2_constrained_smoother`.
* Tuned the new `nav2_constraind_smoother`.

## Description of documentation updates required from your changes

* `nav2_constrained_smoother`'s configure guide will have to be updated with the new parameters.

## Description of how this change was tested

* Copied the current nav2_constrained_smoother as nav2_constrained_smoother_old so that both the improvements can be compared side by side with the older smoother via benchmarking script.
* Made changes to the nav2_constrained_smoother to remove the squaring and added temporary changes to publish the ceres summary data over a topic for becnchmarking.
* Made changes to the smoother benchmarking scripts to be able to incorporate the ceres summaries in results.
* Copied and modified maps for testing from different packages.
* Ran the smoother benchmarking scripts to measure the performance of both smoothers across different maps and with different planners. Tested for 200 trajectories over different maps and planners.
* More detailed results posted on #5072.

---

## Future work that may be required in bullet points

NA

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

### CODE REVIEW

File: nav2_constrained_smoother/README.md [hunk 1]
```diff
@@ -22,8 +22,8 @@ smoother_server:
       minimum_turning_radius: 0.40  # minimum turning radius the robot can perform. Can be set to 0.0 (or w_curve can be set to 0.0 with the same effect) for diff-drive/holonomic robots
       w_curve: 30.0                 # weight to enforce minimum_turning_radius
       w_dist: 0.0                   # weight to bind path to original as optional replacement for cost weight
-      w_smooth: 2000000.0           # weight to maximize smoothness of path
-      w_cost: 0.015                 # weight to steer robot away from collision and cost
+      w_smooth: 5.0e5           # weight to maximize smoothness of path
+      w_cost: 0.12                 # weight to steer robot away from collision and cost
```
Comment (SteveMacenski) [inline nav2_constrained_smoother/README.md]: Please add to the migration guide at docs.nav2.org that this improvement has been had which increases path quality + convergence time. Also update the parameter defaults to be more in line with what folks should use in the configuration guide for the constrained smoother :-)

File: nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp [hunk 2]
```diff
@@ -74,7 +74,7 @@ class SmootherCostFunction
 
   ceres::CostFunction * AutoDiff()
   {
-    return new ceres::AutoDiffCostFunction<SmootherCostFunction, 4, 2, 2, 2>(this);
+    return new ceres::AutoDiffCostFunction<SmootherCostFunction, 6, 2, 2, 2>(this);
```
Comment (SteveMacenski) [inline nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp:77]: For here and elsewhere: can you explain to me the change to `6`?

File: nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp [hunk 3]
```diff
@@ -170,7 +171,7 @@ class SmootherCostFunction
       return;
     }
 
-    r += (T)weight * ki_minus_kmax * ki_minus_kmax;  // objective function value
+    r += (T)std::sqrt(weight) * ki_minus_kmax;  // objective function value
```
Comment (SteveMacenski) [inline nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp]: If we're changing the formulation for the defaults to need adjustment, could the defaults themselves just account on being squared so we don't perform these sqrt operations over the weights? I.e. retune the defaults to be `sqrt` of it's updated value? That would remove some relatively heavy operations 

The cost check points may also be the same?

File: nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp (hunk 2 above)
Comment (videh25) [reply nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp:77]: The template value changed represents the size of residual block returned by this `SmootherCostFunction`. 

The change is made to accommodate for the two additional residual terms, one added for each: `addSmoothResidual` and `addCurvatureResidual`.

Earlier, these were directly returning the squared residual which were of form of  $a^Ta$. However, since we want the squaring to be handled by ceres, now we are passing all the elements of vector $a =[x, y]$ individually for these functions. This caused increase in one residual term for each one of the two mentioned costs and thus, increase of residual terms by 2.

File: nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp (hunk 3 above)
Comment (videh25) [reply nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp]: I too agree with removing the repeated square root functions.
Just listing down ways and concerns in doing so:
1. Square root the weights in configuration and remove `sqrt` from the code.
    - For anyone considering the mathematical formulation of cost function, it is natural to think of the weights mentioned in the config to be the ones directly being multiplied to individual squared costs, since that is how the formulae are generally mentioned in papers and academia. 
    - Weights being squared internally by `ceres` will be misleading.
    - This can work if we also update the documentation to explicitly mention the mathematical formulation used in the smoother to mention that the `yaml` weights are squared by ceres.
2. Keep the `yaml` values and `sqrt` them when the parameter is fetched/updated
    - Values mentioned in the config will be the same as the ones multiplied to the squared cost terms, despite of `ceres` squaring them internally.
    - Might have to rename the class variables to something like `weight_sqrted` since they are not the exact same values as in `yaml`?
    
Please let me know what approach sounds best!

Comment (mini-1235) [reply nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp]: I think the second option probably makes the most sense

Comment (SteveMacenski) [reply nav2_constrained_smoother/include/nav2_constrained_smoother/smoother_cost_function.hpp]: I agree with the second one :-)

**Review by SteveMacenski — APPROVED** [review]
LGTM!

Only 3 asks:
* There are a number of tests that fail now, can you take a look and address? This includes DCO (which is easy if you click on the job it tells you what to run)
* Can you update the migration guide in docs.nav2.org to mention this fix + updated parameters required?
* Does it make sense to update the defaults in the code / example yaml in the configuration guide for the constrained smoother as well to be what you've found works well now?

**Comment by videh25** [comment]
> * There are a number of tests that fail now, can you take a look and address? This includes DCO (which is easy if you click on the job it tells you what to run)
- Still working on the tests. Need to retune the parameters used in the tests for some of them.
- A bunch of them are failing due to improvements in the performance, where smoother is improving the path more than the tests expected. Example: `testingAnchoringToOriginalPath` in [here](https://app.circleci.com/pipelines/github/ros-navigation/navigation2/18258/workflows/f3398768-e404-4d58-8999-8d74fa889ac6/jobs/55036/tests). For such test, should I update the values expected by the tests? Or should I update the tests to a `greater than` comparision rather than the current `expect near`?
> * Can you update the migration guide in docs.nav2.org to mention this fix + updated parameters required?
- Yes, I have those changes ready offline. Will create a PR!
> * Does it make sense to update the defaults in the code / example yaml in the configuration guide for the constrained smoother as well to be what you've found works well now?
- Yes, still finalising the parameters. Will get a final set of parameters once I fix all the tests.

**Comment by mini-1235** [comment]
I would suggest keeping the `EXPECT_NEAR` and updating the expected value. At the same time, please sanity check the new value to make sure it actually makes sense to you

**Comment by SteveMacenski** [comment]
Agreed with @mini-1235

**Comment by SteveMacenski** [comment]
Let me know when you want me to review again :-)

**Comment by videh25** [comment]
Hey @SteveMacenski @mini-1235!

I am almost done with tuning the tests. Just want to run some specific cases through you guys. 

Here is a [google drive folder](https://drive.google.com/drive/folders/1HzJRndiRViJ8BxVffHEr1A0Zf2VRzBo4?usp=drive_link) containing the visuals of how each smoother call in tests improved the path previously vs now. 

During retuning the parameters, I tried to ensure that the qualitative nature of improvements remain the same for each smoother call. However, here are a few notable deviations:
1. `SmootherTest.testingSmoothness` 
    - **Location:** test at `test_constrained_smoother.cpp: 481`
    - This is currently the only failing test.
    - The new `orientation_smoothness_improvement` is lower than the previous expectation.
    - **Old Expected** = 28.5 ±  1.0
    - **New Improvement** = 24.13
    smoothness_test 
    - Increasing `w_smooth` further does not show any significant increase the metric, suggesting that this is not an issue of suboptimal tuning. :-/
    - Need guidance on what to do with this test. Would it be reasonable to change the expected value, given that the qualitative nature of improvement remains the same?
2. `SmootherTest.testingStartGoalOrientations`
    - **Location**: tests at `test_constrained_smoother.cpp: 1059, 1061, 1062`
    - These tests showed a significantly higher improvement in the metric for new smoother.
    - **Old Expected** = 58.9 ±  1.0 
    - **New Improvement** = 99.9
    start_goal_test

    - Even a small value of `w_smooth` results in a similar improvement (tried tuning it down, does not work).
    - Considering the parameters for this smoother call, it makes sense to observe the new behaviour. (Only `w_smooth` and `w_curve` are non-zero --> Path straightens out prioritizing only smoothness).
    - Start and goal orientations also change to point along the line connecting them. Thus, both having an orientation of `PI/4` rather than the previously hard coded `PI/8` and `7*PI/8`.
    - Have updated the tests to expect the new path.
3. `SmootherTest.testingObstacleAvoidanceNearCusps`
    - New improvements are in general more aggressive compared to the previous improvements. 
    - Please take a look at the results of smoother calls in this [folder](https://drive.google.com/drive/folders/1426FXJBtVsnNBh_vzDLrl2EGJJ8K21uo?usp=drive_link).

**Comment by SteveMacenski** [comment]
> Would it be reasonable to change the expected value, given that the qualitative nature of improvement remains the same?

Yes, this is a different formulation so I would expect some variation. They both look good though :+1: 

> Considering the parameters for this smoother call, it makes sense to observe the new behaviour. (Only w_smooth and w_curve are non-zero --> Path straightens out prioritizing only smoothness).

Does this test not try to maintain the start/goal orientations? If so, then this makes sense. Technically a straight line is the lowest-curvature option as long as the boundary conditions aren't being respected to force a curve.

>  Please take a look at the results of smoother calls in this [folder](https://drive.google.com/drive/folders/1426FXJBtVsnNBh_vzDLrl2EGJJ8K21uo?usp=drive_link).

A number of these look rather odd, having pretty extreme moves (like #4) that don't seem justified / well traded off for smoothness / length. But I suppose the originals also look a little extreme. Can we tone that down a bit? The behavior in #2 looks pretty reasonable.

**Comment by videh25** [comment]
> Does this test not try to maintain the start/goal orientations? If so, then this makes sense. Technically a straight line is the lowest-curvature option as long as the boundary conditions aren't being respected to force a curve.

No, that test is specifically for the case where start and goal locations are not maintained.

> A number of these look rather odd, having pretty extreme moves (like https://github.com/ros-navigation/navigation2/pull/4) that don't seem justified / well traded off for smoothness / length. But I suppose the originals also look a little extreme. Can we tone that down a bit? The behavior in https://github.com/ros-navigation/navigation2/pull/2 looks pretty reasonable.

I have tuned the parameters in `testingObstacleAvoidanceNearCusps` to generate smoother paths. Please find the new images in the [same folder](https://drive.google.com/drive/folders/1426FXJBtVsnNBh_vzDLrl2EGJJ8K21uo?usp=drive_link). 

However, a test is failing now. Here is the context for that:
- Step1: Smooth with baseline params.
- Step2: Smooth with lower w_cost and higher w_cost_cusp_multiplier (less penalty on normal poses, more on cusps)
- The test expects `origin_similarity_improvement > 0`, assuming Step2 stays closer to the original path due to lower `w_cost`.
- The test expects `origin_similarity_improvement` to be positive. It assumes the path form Step2 will be closer to original path due to smaller `w_cost`.
- In practice, increasing w_cost_cusp_multiplier shifts cusps farther from obstacles. Since current paths are prioritizing smoothness more, the deviation in Step2 path is more than the deviation in Step1, so the metric becomes negative.
- Need advice on what to do about these tests.

Also, `testingObstacleAvoidanceNearCusps` contains multiple inter-dependent `EXPECT` statements which is makes the parameter tuning for the test brittle. Changes that satisfy one assertion often break another. I’ve tried multiple parameter combinations to avoid the current failure, but they consistently violate at least one of the other expectations. This is the best trade-off scenario I was able to achieve among all those tuning trials.

**Comment by SteveMacenski** [comment]
> I have tuned the parameters in testingObstacleAvoidanceNearCusps to generate smoother paths. Please find the new images in the [same folder](https://drive.google.com/drive/folders/1426FXJBtVsnNBh_vzDLrl2EGJJ8K21uo?usp=drive_link).

Looks great to me!

For the tests: I'd say delete any that don't pass your sniff test and create some new ones that are better / less fragile. The cusp movement feature was added after the fact and may have been too prescriptive based on the behavior at the time rather than good general-purpose tests (it appears)

**Comment by videh25** [comment]
Here is a summary of changes made in tests in `testingObstacleAvoidanceNearCusps` for your reference:
- Introduced a `cusp_cost_criterion`, which considers solely the cost at the cusp pose.
- Replaced all worst-cost comparisions with cusp-cost comparisions, since cusp cost exhibits more predictable behaviour, whereas worst-cost pose varies with path geometry.
- Removed origin similarity test due to it being sensitive to other parameters and can change with tuning.
- Removed a repeated checks .(previously at 830-844 in the older version, already covered earlier in the test).

**Comment by videh25** [comment]
Opened a corresponding PR in `nav2_docs` [here](https://github.com/ros-navigation/docs.nav2.org/pull/895).

**Review by SteveMacenski — APPROVED** [review]
